3rdParty/rocksdb/v5.1.4/include/rocksdb/env_encryption.h

```diff
@@ -125,27 +132,65 @@ class EncryptionProvider {
  public:
     virtual ~EncryptionProvider() {};
 
+    // GetPrefixLength returns the length of the prefix that is added to every file
+    // and used for storing encryption options.
+    // For optimal performance, the prefix length should be a multiple of 
+    // the a page size.
+    virtual size_t GetPrefixLength() = 0;
+
+    // CreateNewPrefix initialized an allocated block of prefix memory 
+    // for a new file.
+    virtual Status CreateNewPrefix(const std::string& fname, char *prefix, size_t prefixLength) = 0;
+
     // CreateCipherStream creates a block access cipher stream for a file given
     // given name and options.
     virtual Status CreateCipherStream(const std::string& fname, const EnvOptions& options,
-                                 unique_ptr<BlockAccessCipherStream>* result) = 0;
+      Slice& prefix, unique_ptr<BlockAccessCipherStream>* result) = 0;
 };
 
 // This encryption provider uses a CTR cipher stream, with a given block cipher 
 // and IV.
+//
+// Note: This is a possible implementation of EncryptionProvider, 
+// it is considered suitable for use, provided a safe BlockCipher is used.
 class CTREncryptionProvider : public EncryptionProvider {
     private:
       BlockCipher& cipher_;
-      const char* iv_;
+    protected:
+      const static size_t defaultPrefixLength = 4096;
```

Comment:
I'm not too worried about people screwing it up. To _actually_ screw it up they'd have to set it to something less than 32B, right? They'd have huge issues with RocksDB in general if they did that. A larger block doesn't hurt anything other than a little more wasted space.

I think there are potentially benefits to keeping the prefix block the same as the RocksDB block size, since it will keep the access patterns uniform. Otherwise we end up with something like a 4KB block at the beginning and 16KB or 32KB blocks after that. Right? It could potentially throw off alignment, etc.